libs/payments/cart/src/lib/cart.service.ts

```diff
@@ -92,16 +96,98 @@ export class CartService {
       if (error instanceof CartNotUpdatedError) {
         errorReasonId = CartErrorReasonId.BASIC_ERROR;
       }
-
-      await this.cartManager
-        .finishErrorCart(cartId, {
+      try {
+        await this.cartManager.finishErrorCart(cartId, {
           errorReasonId,
-        })
-        .catch((e) => {
-          // We silently fail here so as not to eat the original error
-          Sentry.captureException(e);
         });
 
+        const cart = await this.cartManager.fetchCartById(cartId);
+        if (cart.stripeSubscriptionId) {
+          const subscription = await this.subscriptionManager.retrieve(
+            cart.stripeSubscriptionId
+          );
+          const invoice = subscription.latest_invoice
+            ? await this.invoiceManager.retrieve(subscription.latest_invoice)
+            : undefined;
+          if (invoice) {
+            switch (invoice.status) {
+              case 'draft':
+                await this.invoiceManager.delete(invoice.id);
+                break;
+              case 'open':
+              case 'uncollectible':
+                await this.invoiceManager.void(invoice.id);
+                break;
+              case 'paid':
+                throw new CartError('Paid invoice found on failed cart', {
+                  cartId,
+                  stripeCustomerId: cart.stripeCustomerId,
+                  invoiceId: invoice.id,
+                });
+            }
+          }
+
+          const paymentIntent =
+            await this.subscriptionManager.getLatestPaymentIntent(subscription);
+          if (paymentIntent) {
+            switch (paymentIntent.status) {
+              case 'requires_payment_method':
+              case 'requires_capture':
+              case 'requires_confirmation':
+              case 'requires_action':
+              case 'processing':
+                await this.paymentIntentManager.cancel(paymentIntent.id);
```

Comment:
Does Stripe throw an error if we call `cancel` on a succeeded payment intent?

There's a non-zero chance given the async timing here that the payment intent moves from processing to succeeded in the time between getLatestPaymentIntent and cancel.

Reply:
Presumably, yes. [From their docs](https://docs.stripe.com/payments/paymentintents/lifecycle) "You can cancel a PaymentIntent at any point before it’s in a `processing` or `succeeded` state". Because of the way we are logging any errors thrown in this block to sentry, I think its alright to have an unhandled exception here. It won't overwrite the original error, and should get logged successfully.

Reply:
from call: this switch/case is not needed broadly, because any failed cancellation will trigger a sentry alert